src/core/fetchers/segment/prioritizer.ts

```diff
@@ -136,87 +223,243 @@ export default class ObservablePrioritizer<T> {
+        }
+        this._onTaskBegin(newTask);
+        return observableConcat(
+          obs.pipe(map((data) => ({ type: "data" as const, value: data }))),
+          observableOf({ type: "ended" as const })
+        ).pipe(finalize(() => {
+          this._onTaskEnd(newTask);
+        }));
+      };
+
+      if (!this._canBeStartedNow(newTask)) {
+        // This task doesn't have priority yet. Start it on trigger
+        this._waitingQueue.push(newTask);
+        return trigger.pipe(switchMap(onTrigger));
       }
+
+      // Start it right away
+      const startTask$ = trigger.pipe(startWith(true), switchMap(onTrigger));
+      return !this._isHighPriority(newTask) ?
+        startTask$ :
+        observableMerge(startTask$,
+
+                        // Note: we want to begin interrupting low-priority tasks just
+                        // after starting the current one because the interrupting
+                        // logic can call external code.
+                        // This would mean re-entrancy, itself meaning that some weird
+                        // half-state could be reached unless we're very careful.
+                        // To be sure no harm is done, we put that code at the last
+                        // possible position (the previous Observable sould be
+                        // performing all its initialization synchronously).
+                        observableDefer(() => { this._interruptCancellableTasks();
+                                                return EMPTY; }));
     });
+
     return pObs$;
   }
 
   /**
-   * Update the priority of an Observable created through the create method.
-   *
-   * Note that this will only have an effect on Observable which are not yet
-   * started.
-   * This means it will only have an effect on:
-   *   - unsubscribed Observables
-   *   - Observables waiting for Observables with an higher priority to
-   *     finish
-   *
+   * Update the priority of an Observable created through the `create` method.
    * @param {Observable} obs
    * @param {number} priority
    */
-  public updatePriority(obs : Observable<T>, priority : number) : void {
-    const index = arrayFindIndex(this._queue,
-                                 (elt) => elt.observable === obs);
+  public updatePriority(obs : Observable<ITaskEvent<T>>, priority : number) : void {
+    const waitingQueueIndex = arrayFindIndex(this._waitingQueue,
+                                             (elt) => elt.observable === obs);
+
+    if (waitingQueueIndex >= 0) { // If it was still waiting for its turn
+      const waitingQueueElt = this._waitingQueue[waitingQueueIndex];
+      if (waitingQueueElt.priority === priority) {
+        return;
+      }
+
+      waitingQueueElt.priority = priority;
+
+      if (!this._canBeStartedNow(waitingQueueElt)) {
+        return;
+      }
+
+      this._startTask(waitingQueueIndex);
+
+      if (this._isHighPriority(waitingQueueElt)) {
```

Comment:
I suppose it is volontary here to cancel tasks after having started a new one. Maybe we could add a comment so that it would be more explicit that it was made on purpose.

Reply:
Done